modules: add support for conflict in lua modulefile

xdelaruelle opened the pull request:
Add support for conflict directives in Lua modulefile like done for Tcl modulefile.

Note that conflicts are correctly honored on Lmod and Environment Modules <4.2 only if mutually expressed on both modulefiles that conflict with each other.

Fixes #21987.

xdelaruelle committed:
modules: add support for conflict in lua modulefile

Add support for conflict directives in Lua modulefile like done for Tcl
modulefile.

Note that conflicts are correctly honored on Lmod and Environment
Modules <4.2 only if mutually expressed on both modulefiles that
conflict with each other.

Migrate conflict code from Tcl-specific classes to the common part. Add
tests for Lmod and split the conflict test case in two.

Fixes #21987.

alalazo commented:
@xdelaruelle Can you solve the conflict here? I'll review asap after that.

xdelaruelle commented:
> @xdelaruelle Can you solve the conflict here? I'll review asap after that.

@alalazo Merge conflict fixed

alalazo commented on lib/spack/spack/modules/common.py:
```diff
@@ -763,6 +769,29 @@ def has_manpath_modifications(self):
+        projection = proj.get_projection(self.conf.projections, self.spec)
+        f = string.Formatter()
+        for item in self.conf.conflicts:
+            if len([x for x in f.parse(item)]) > 1:
+                for naming_dir, conflict_dir in zip(projection.split("/"), item.split("/")):
+                    if naming_dir != conflict_dir:
+                        message = "conflict scheme does not match naming "
+                        message += "scheme [{spec}]\n\n"
+                        message += 'naming scheme   : "{nformat}"\n'
+                        message += 'conflict scheme : "{cformat}"\n\n'
+                        message += "** You may want to check your "
+                        message += "`modules.yaml` configuration file **\n"
+                        tty.error(message.format(spec=self.spec, nformat=projection, cformat=item))
+                        raise SystemExit("Module generation aborted.")
```
This exception is definitely too low level (it's not caught even by `Exception`). Can we have an ad-hoc exception that inherits from `SpackError`?

alalazo replied:
This can be extracted as a separate function, with a more meaningful name. Maybe:
```python
self._verify_conflict_naming_consistency_or_raise(item)
```
?

alalazo commented on lib/spack/spack/modules/common.py:
```diff
@@ -763,6 +769,29 @@ def has_manpath_modifications(self):
+                        message += "`modules.yaml` configuration file **\n"
+                        tty.error(message.format(spec=self.spec, nformat=projection, cformat=item))
+                        raise SystemExit("Module generation aborted.")
+                item = self.spec.format(item)
+            fmts.append(item)
+        # Substitute spec tokens if present
+        return [self.spec.format(x) for x in fmts]
```
Why do we need to pass items through `self.spec.format` two times?

alalazo commented on lib/spack/spack/modules/common.py:
```diff
@@ -470,6 +471,11 @@ def hash(self):
             return self.spec.dag_hash(length=hash_length)
         return None
 
+    @property
+    def conflicts(self):
+        """Conflicts for this module file"""
+        return self.conf.get("conflict", [])
```
This is fine, but now we can also remove the same property in `TclConfiguration`.

xdelaruelle replied:
done

alalazo commented on lib/spack/spack/test/modules/lmod.py:
```diff
@@ -290,6 +290,23 @@ def test_non_virtual_in_hierarchy(self, factory, module_configuration):
         with pytest.raises(spack.modules.lmod.NonVirtualInHierarchyError):
             module.write()
 
+    def test_conflicts(self, modulefile_content, module_configuration):
+        """Tests adding conflicts to the module."""
+
+        # This configuration has no error, so check the conflicts directives
+        # are there
+        module_configuration("conflicts")
+        content = modulefile_content("mpileaks")
+
+        assert len([x for x in content if x.startswith("conflict")]) == 2
+        assert len([x for x in content if x == 'conflict("mpileaks")']) == 1
+        assert len([x for x in content if x == 'conflict("intel/14.0.1")']) == 1
+
+        # This configuration is inconsistent, check an error is raised
+        module_configuration("wrong_conflicts")
+        with pytest.raises(SystemExit):
+            modulefile_content("mpileaks")
```
These are two tests, can we split them?

xdelaruelle replied:
done

alalazo reviewed (changes requested):
I have a few minor requests, but this looks great. Thanks!

xdelaruelle commented:
@alalazo I will look at your change requests but I may not have an answer for everything as my work here was to copy/paste the conflict support code already available for tcl modulefiles.

xdelaruelle commented:
> I can try to help, or push directly to the branch, if you're fine with that.

@alalazo I have handled some of your review comments and it would be great if you can help with the 3 comments relative to the `conflicts` method code refactoring. Please push directly to the branch, I am fine with that.

alalazo committed:
modules: minor refactor of conflicts

In case of inconsistency raise ModulesError
instead of SystemError.

Extract function to verify naming consistency

Remove double formatting

alalazo commented:
@xdelaruelle Pushed a commit. If you can double check it, I'll merge as soon as CI is :heavy_check_mark:

xdelaruelle commented on lib/spack/spack/modules/common.py:
```diff
@@ -774,23 +774,29 @@ def conflicts(self):
+                if naming_dir != conflict_dir:
+                    errors.extend(
+                        [
+                            f"spec={self.spec.cshort_spec}" f"conflict_scheme={item}",
```
Isn't there a missing `,` between f"spec*" and f"conflict_scheme*"?

Error is hard to read otherwise:

```
E           spack.modules.common.ModulesError: conflict scheme does not match naming scheme
E               spec=mpileaks@2.3%gcc@10.2.1~debug~opt+shared+static build_system=generic arch=test-debian6-core2/be7geaaconflict_scheme={name}/{compiler.name}
E           naming_scheme={name}/{version}-{compiler.name}
```

alalazo replied:
There is indeed!

alalazo replied:
Fixed

alalazo committed:
Add missing comma in literal list

xdelaruelle commented on lib/spack/spack/modules/common.py:
```diff
@@ -763,6 +769,36 @@ def has_manpath_modifications(self):
+        if errors:
+            raise ModulesError(
+                message="conflict scheme does not match naming scheme",
+                long_message="\n".join(errors),
```
LGTM. I may suggest however a last change to improve the layout of the error message:
```python
                long_message="\n    ".join(errors),
```

Message format will be improved this way, for instance:

```
E           spack.modules.common.ModulesError: conflict scheme does not match naming scheme
E               spec=mpileaks@2.3%gcc@10.2.1~debug~opt+shared+static build_system=generic arch=test-debian6-core2/be7geaa
E               conflict_scheme={name}/{compiler.name}
E               naming_scheme={name}/{version}-{compiler.name}
```

Instead of:
```
E           spack.modules.common.ModulesError: conflict scheme does not match naming scheme
E               spec=mpileaks@2.3%gcc@10.2.1~debug~opt+shared+static build_system=generic arch=test-debian6-core2/be7geaa
E           conflict_scheme={name}/{compiler.name}
E           naming_scheme={name}/{version}-{compiler.name}
```

alalazo replied:
Done, but we need to revisit this in another PR. Since we don't know how and where the error message is printed it would be better that the base error message class takes a list of additional lines without space added where the exception is raised.

alalazo commented on lib/spack/spack/modules/common.py:
```diff
@@ -763,6 +769,36 @@ def has_manpath_modifications(self):
+        if errors:
+            raise ModulesError(
+                message="conflict scheme does not match naming scheme",
+                long_message="\n".join(errors),
```
```suggestion
                long_message="\n    ".join(errors),
```

alalazo committed:
Improve layout

alalazo commented:
Thanks @xdelaruelle !
